qcodes/instrument/parameter.py

```diff
@@ -1745,15 +1730,174 @@ def setpoint_full_names(self) -> Optional[Sequence[Sequence[str]]]:
+        the parameter (that is available when calling ``cache.get()`` or
+        ``get_latest()``), and will NOT pass that value to the instrument.
+
+        Note that this method also respects all the validation, parsing,
+        offsetting, etc that the parameter's ``set`` method respects. However,
+        if the parameter has :attr:`step` defined, unlike the ``set`` method,
+        this method does not perform setting the parameter step-by-step.
+
+        Args:
+            value: new value for the parameter
+        """
+        self._parameter.validate(value)
+        raw_value = self._parameter._from_value_to_raw_value(value)
+        self._update_with(value=value, raw_value=raw_value)
+
+    def _update_with(self, *,
+                     value: ParamDataType,
+                     raw_value: ParamRawDataType,
+                     timestamp: Optional[datetime] = None
+                     ) -> None:
+        """
+        Simply overwrites the value, raw value, and timestamp in this cache
+        with new ones.
+
+        Args:
+            value: new value of the parameter
+            raw_value: new raw value of the parameter
+            timestamp: new timestamp of the parameter; if ``None``,
+                then timestamp of "now" is used
+        """
+        self._value = value
+        self._raw_value = raw_value
+        if timestamp is None:
+            self._timestamp = datetime.now()
+        else:
+            self._timestamp = timestamp
+
+    def get(self, get_if_invalid: bool = True) -> ParamDataType:
+        """
+        Return cached value if time since get was less than ``max_val_age``,
+        otherwise perform ``get()`` on the parameter and return result. A
+        ``get()`` will also be performed if the parameter has never been
+        captured but only if ``get_if_invalid`` argument is ``True``.
+
+        Args:
+            get_if_invalid: if set to ``True``, ``get()`` on a parameter
+                will be performed in case the cached value is invalid (for
+                example, due to ``max_val_age``, or because the parameter has
+                never been captured)
+        """
+        no_get = not hasattr(self._parameter, 'get')
+
+        # the parameter has never been captured so `get` it but only
+        # if `get_if_invalid` is True
+        if self._timestamp is None:
+            if get_if_invalid:
+                if no_get:
+                    raise RuntimeError(f"Value of parameter "
+                                       f"{(self._parameter.full_name)} "
+                                       f"is unknown and the Parameter does "
+                                       f"not have a get command. "
+                                       f"Please set the value before "
+                                       f"attempting to get it.")
+                return self._parameter.get()
+            else:
+                return self._value
```

Comment:
Is there any chance that _value is not `None` here?

Reply:
yes. especially when "None" is not a valid value for a parameter.

i'm not sure i understand the question - what did you mean to wonder?

Reply:
ah, i get it. I think this will likely be rewised in the snapshot_base-improvement-related PR later. also extra tests for _Cache in the not-so-far-future would help.